plumbing/format/pktline/pktline.go

```diff
@@ -0,0 +1,229 @@
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: > 0000")
+		}
+	}()
+
+	_, err = w.Write(flushPkt)
+	return err
+}
+
+// WriteDelim writes a delimiter packet.
+// This always writes 4 bytes.
+func WriteDelim(w io.Writer) (err error) {
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: > 0001")
+		}
+	}()
+
+	_, err = w.Write(delimPkt)
+	return err
+}
+
+// WriteResponseEnd writes a response-end packet.
+// This always writes 4 bytes.
+func WriteResponseEnd(w io.Writer) (err error) {
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: > 0002")
+		}
+	}()
+
+	_, err = w.Write(responseEndPkt)
+	return err
+}
+
+// ReadPacket reads a pktline packet.
+// This returns the length of the packet, the packet payload, and an error.
+// The error can be of type *ErrorLine if the packet is an error packet.
+// Use packet length to determine the type of packet i.e. 0 is a flush packet,
+// 1 is a delim packet, 2 is a response-end packet, and a length greater or
+// equal to 4 is a data packet.
+func ReadPacket(r io.Reader) (l int, p []byte, err error) {
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: < %04x %s", l, p)
+		}
+	}()
+
+	var pktlen [lenSize]byte
+	n, err := io.ReadFull(r, pktlen[:])
+	if err != nil {
+		if errors.Is(err, io.ErrUnexpectedEOF) {
+			return Err, nil, fmt.Errorf("%w: %d", ErrInvalidPktLen, n)
+		}
+
+		return Err, nil, err
+	}
+
+	if n != lenSize {
+		return Err, nil, fmt.Errorf("%w: %d", ErrInvalidPktLen, n)
+	}
+
+	length, err := ParseLength(pktlen[:])
+	if err != nil {
+		return Err, nil, err
+	}
+
+	switch length {
+	case Flush, Delim, ResponseEnd:
+		return length, nil, nil
+	case lenSize: // empty line
+		return length, []byte{}, nil
+	}
+
+	dataLen := length - lenSize
+	data := make([]byte, 0, dataLen)
+	dn, err := io.ReadFull(r, data[:dataLen])
+	if err != nil {
+		return Err, nil, err
+	}
+
+	if dn != dataLen {
+		return Err, data, fmt.Errorf("%w: %d", ErrInvalidPktLen, dn)
+	}
+
+	buf := data[:dn]
+	if bytes.HasPrefix(buf, errPrefix) {
+		err = &ErrorLine{
+			Text: string(bytes.TrimSpace(buf[4:])),
+		}
+	}
+
+	return length, buf, err
+}
+
+// ReadPacketString reads a pktline packet and returns it as a string.
+// The returned string is trimmed of whitespace.
+func ReadPacketString(r io.Reader) (l int, s string, err error) {
+	l, p, err := ReadPacket(r)
+	return l, string(bytes.TrimSpace(p)), err
+}
+
+// PeekPacket reads a pktline packet without consuming it.
+// This returns the length of the packet, the packet payload, and an error.
+// The error can be of type *ErrorLine if the packet is an error packet.
+// Use packet length to determine the type of packet i.e. 0 is a flush packet,
+// 1 is a delim packet, 2 is a response-end packet, and a length greater or
+// equal to 4 is a data packet.
+func PeekPacket(r ioutil.ReadPeeker) (l int, p []byte, err error) {
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: < %04x %s", l, p)
```

Comment:
This is adding an additional allocation even when trace is disabled - which is quite costly.
Just for testing purposes, I made this changes which resulted in no additional alloc:

```suggestion
			trace.Packet.Printfunc(func() string { return fmt.Sprintf("packet: < %04x %s", l, p) })
```
And adding on `trace`:
```go
func (t Target) Printfunc(data func() string) {
	if int32(t)&current.Load() != 0 {
		logger.Output(2, data()) // nolint: errcheck
	}
}
```

We may need to think a bit more on ways around this to keep a nice API while dodging this additional allocation.